### ISSUE TITLE
gradient checkpointing with use_reentrant=False cannot reduce peak memory

### ISSUE THREAD

**Issue by KK666-AI** [pytorch/pytorch#147449]
### 🐛 Describe the bug

I am comparing the memory cost between `use_reentrant=False` and `use_reentrant=True` when using gradient checkpointing. When set `use_reentrant=False`, i find the peak memory is exactly the same with the one without using gradient checkpointing.

The following is my script, suppose the script name is `test.py`
```

import argparse
import torch
import torch.nn as nn
import GPUtil

def get_gpu_memory():
    memory_dict = dict()
    gpus = GPUtil.getGPUs()
    for gpu in gpus:
        total = gpu.memoryTotal / 1024  # GB
        used = gpu.memoryUsed / 1024 # GB
        free = gpu.memoryFree / 1024  # GB
        memory = {
            "total": total,
            "used": used,
            "free": free,
            "used_ratio": used/total
        }
        memory_dict[gpu.id] = memory

    return memory_dict


class CustomLayer(nn.Module):
    def __init__(self, d_model, nhead):
        super(CustomLayer, self).__init__()
        self.self_attn = nn.MultiheadAttention(d_model, nhead)
        self.linear1 = nn.Linear(d_model, d_model * 4)
        self.linear2 = nn.Linear(d_model * 4, d_model)
        self.norm1 = nn.LayerNorm(d_model)
        self.norm2 = nn.LayerNorm(d_model)
        self.dropout = nn.Dropout(0.1)

    def forward(self, x):
        # Self-attention
        attn_output, _ = self.self_attn(x, x, x)
        x = self.norm1(x + attn_output)  # Residual connection + normalization

        # Feedforward network
        ff_output = self.linear1(x)
        ff_output = torch.relu(ff_output)
        ff_output = self.linear2(ff_output)
        x = self.norm2(x + ff_output)  # Residual connection + normalization
        x = self.dropout(x)  # test dropout.

        return x

class Model(nn.Module):
    def __init__(self, d_model, nhead, num_layers, use_checkpointing=False, use_reentrant=False):
        super(Model, self).__init__()
        self.linear = nn.Linear(d_model, d_model)
        self.layers = nn.ModuleList([CustomLayer(d_model, nhead) for _ in range(num_layers)])
        self.final_layer = nn.Linear(d_model, 1)  # Example output layer
        self.use_checkpointing = use_checkpointing
        self.use_reentrant = use_reentrant

    def _forward_layers(self, layer_group, x):
        for layer in layer_group:
            x = layer(x)
        return x

    def native_forward(self, x):
        for layer in self.layers:
            x = layer(x)

        return x

    def checkpointing_group_layers_forward(self, x):
        interval = 4
        for i in range(0, len(self.layers), interval):
            layer_group = self.layers[i:min(i + interval, len(self.layers))]  # Get a group of two layers
            x = torch.utils.checkpoint.checkpoint(lambda x: self._forward_layers(layer_group, x), x, preserve_rng_state=False, use_reentrant=self.use_reentrant)
        return x

    def forward(self, x):
        if self.use_checkpointing:
            x = self.checkpointing_group_layers_forward(x)
        else:
            x = self.native_forward(x)

        return self.final_layer(x)


def train(model, dataloader, optimizer, device, epoch):
    model.train()
    for epoch in range(epoch):  # Number of epochs
        for batch_idx, (data, target) in enumerate(dataloader):
            data, target = data.to(device), target.to(device)
            optimizer.zero_grad()
            output = model(data)
            loss = nn.MSELoss()(output, target.view(-1, 1))  # Assuming target has a shape compatible with output
            loss.backward()  # Backward pass
            optimizer.step()  # Update model parameters

            if batch_idx % 10 == 0:
                memory_dict = get_gpu_memory()[0]  # device 0
                total = memory_dict["total"]
                used = memory_dict["used"]
                used_ratio = memory_dict["used_ratio"]
                print(f"Epoch: {epoch}, Batch: {batch_idx}, Loss: {loss.item()}, "
                            f"total={'%.2f' % total}, used={'%.2f' % used}, used_ratio={'%.2f' % used_ratio}")

def main(args):
    data = torch.randn(args.seq_len, args.dim)  # (sequence_length, dim)
    target = torch.randn(args.seq_len, 1)  # Assume a regression task

    dataset = torch.utils.data.TensorDataset(data, target)
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=args.batch_size, shuffle=True, drop_last=True)

    # Initialize the model, optimizer, and move to device
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    model = Model(args.dim, args.nhead, args.num_layers, args.use_checkpointing, args.use_reentrant).to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

    # Train the model
    train(model, dataloader, optimizer, device, args.epoch)

    print("end.")


def get_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--seq_len",
        type=int,
        default=2000,
        help="sequence length",
    )
    parser.add_argument(
        "--dim",
        type=int,
        default=1024 * 2,
        help="dimension",
    )
    parser.add_argument(
        "--nhead",
        type=int,
        default=4,
        help="number of head",
    )
    parser.add_argument(
        "--num_layers",
        type=int,
        default=64,
        help="number of layers",
    )
    parser.add_argument(
        "--batch_size",
        type=int,
        default=12,
        help="batch size",
    )
    parser.add_argument(
        "--epoch",
        type=int,
        default=2,
        help="epoch",
    )
    parser.add_argument(
        "--use_checkpointing",
        action="store_true",
        default=False,
        help="whether use checkpointing",
    )
    parser.add_argument(
        "--use_reentrant",
        action="store_true",
        default=False,
        help="whether use reentrant",
    )

    args = parser.parse_args()

    return args


if __name__ == "__main__":
    args = get_args()
    main(args)
```

I run three experients:

1) run experient without gradient checkpointing `python -u test.py` , the peak memory cost is 60.90GB.

```
Epoch: 0, Batch: 0, Loss: 1.738878607749939, total=79.65, used=60.89, used_ratio=0.76
Epoch: 0, Batch: 10, Loss: 2.8586766719818115, total=79.65, used=60.90, used_ratio=0.76
Epoch: 0, Batch: 20, Loss: 1.9227111339569092, total=79.65, used=60.90, used_ratio=0.76
```

2) run experient with gradient checkpointing, but set `use_reentrant=False`, `python -u test.py --use_checkpointing`, the peak memory cost is 60.84GB

```
Epoch: 0, Batch: 0, Loss: 1.4675014019012451, total=79.65, used=60.84, used_ratio=0.76
Epoch: 0, Batch: 10, Loss: 1.1877995729446411, total=79.65, used=60.84, used_ratio=0.76
Epoch: 0, Batch: 20, Loss: 1.4552075862884521, total=79.65, used=60.84, used_ratio=0.76

```

3) run experient with gradient checkpointing, but set `use_reentrant=True`, `python -u test.py --use_checkpointing --use_reentrant`, the peak memory cost is 12.82GB

```
/opt/conda/envs/folding/lib/python3.11/site-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
Epoch: 0, Batch: 0, Loss: 1.1766417026519775, total=79.65, used=12.82, used_ratio=0.16
/opt/conda/envs/folding/lib/python3.11/site-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
Epoch: 0, Batch: 10, Loss: 0.6375870704650879, total=79.65, used=12.82, used_ratio=0.16
/opt/conda/envs/folding/lib/python3.11/site-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
Epoch: 0, Batch: 20, Loss: 2.0402159690856934, total=79.65, used=12.82, used_ratio=0.16
/opt/conda/envs/folding/lib/python3.11/site-packages/torch/utils/checkpoint.py:87: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
  warnings.warn(
```

### Versions

Collecting environment information...
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.5
Libc version: glibc-2.35

Python version: 3.11.10 | packaged by conda-forge | (main, Oct 16 2024, 01:27:36) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.1.0-23-cloud-amd64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 550.90.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               208
On-line CPU(s) list:                  0-207
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8481C CPU @ 2.70GHz
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   52
Socket(s):                            2
Stepping:                             8
BogoMIPS:                             5399.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid cldemote movdiri movdir64b fsrm md_clear serialize amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            4.9 MiB (104 instances)
L1i cache:                            3.3 MiB (104 instances)
L2 cache:                             208 MiB (104 instances)
L3 cache:                             210 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-51,104-155
NUMA node1 CPU(s):                    52-103,156-207
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.0
[pip3] torch==2.5.1+cu124
[pip3] torchaudio==2.5.1+cu124
[pip3] torchelastic==0.2.2
[pip3] torchvision==0.20.1+cu124
[pip3] triton==3.1.0
[conda] numpy                     2.1.2           py311h71ddf71_0    conda-forge
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.5.1+cu124              pypi_0    pypi
[conda] torchaudio                2.5.1+cu124              pypi_0    pypi
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchvision               0.20.1+cu124             pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi

**Comment by yanty123** [comment]
same question

**Comment by AbnerVictor** [comment]
Same issue here, waiting for a solution
